test/unittests/instructions_test.cpp

```diff
@@ -16,36 +20,78 @@ constexpr int get_revision_defined_in(size_t op) noexcept
 {
     for (size_t r = EVMC_FRONTIER; r <= EVMC_MAX_REVISION; ++r)
     {
-        if (evmone::instr::gas_costs[r][op] != evmone::instr::undefined)
+        if (instr::gas_costs[r][op] != instr::undefined)
             return static_cast<int>(r);
     }
     return unspecified;
 }
-}  // namespace
 
-TEST(instructions, validate_since)
+constexpr bool is_terminating(evmc_opcode op) noexcept
 {
-    for (size_t op = 0x00; op <= 0xff; ++op)
+    switch (op)
     {
-        const auto since = evmone::instr::traits[op].since;
-        const auto test_value = since.has_value() ? *since : unspecified;
-        const auto expected = get_revision_defined_in(op);
-        EXPECT_EQ(test_value, expected) << std::hex << op;
+    case OP_STOP:
+    case OP_RETURN:
+    case OP_REVERT:
+    case OP_INVALID:
+    case OP_SELFDESTRUCT:
+        return true;
+    default:
+        return false;
     }
 }
 
+template <evmc_opcode Op>
+constexpr void validate_traits_of() noexcept
+{
+    constexpr auto tr = instr::traits[Op];
+
+    // immediate_size
+    if constexpr (Op >= OP_PUSH1 && Op <= OP_PUSH32)
+        static_assert(tr.immediate_size == Op - OP_PUSH1 + 1);
+    else
+        static_assert(tr.immediate_size == 0);
+
+    // is_terminating
+    static_assert(tr.is_terminating == is_terminating(Op));
+    static_assert(!tr.is_terminating || tr.immediate_size == 0);
+
+    // since
+    constexpr auto expected_rev = get_revision_defined_in(Op);
+    static_assert(tr.since.has_value() ? *tr.since == expected_rev : expected_rev == unspecified);
+}
+
+template <std::size_t... Ops>
+constexpr auto validate_traits(std::index_sequence<Ops...>)
+{
+    return (validate_traits_of<static_cast<evmc_opcode>(Ops)>(), ..., true);
```

Comment:
this is confusing... what type does `validate_traits` return?

Reply:
ah, I think I got, it always returns `true` as result of repeated `,` operator, but calls 256 `validate_traits_of` as "side-effect" of expression.

Reply:
maybe possible to split to two lines
```
validate_traits_of<static_cast<evmc_opcode>(Ops)>(), ...;
return true;
```
?